[CORE-14511] [CORE-14518] dt/cluster_linking: Wait for topic and partitions

BenPope opened the pull request:
If the topic is going to be produced to, wait for the partitions, too.

For example, this avoids the KgoVerifier warning:
```
"More partitions in valid_offsets file than in topic!"
```

E.g.: https://ci-artifacts.dev.vectorized.cloud/production-devprod/redpanda/75412/019a3a6a-4e02-4342-9b24-7ecdaa64c418/vbuild/ducktape/results/2025-10-31--001/report.html

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

oleiman reviewed (dismissed):
nice one

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75427-->
<!--f719bf4c-c005-4489-8b17-b75102f64b9a-->
<!--bbd38ab8-6aa0-43ad-8c04-3baa5365c43c-->
<!--title-->
## Retry command for [Build#75427](https://buildkite.com/redpanda/redpanda/builds/75427)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkUpdateBrokersTests.test_update_brokers
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--fa6dcb3c-617f-4d7f-9280-8ffd1e8fad21-->
<!--TEST RESULTS-->
<!--3d8c735a-4bc5-4bcc-8663-a69def560ffc-->
<!--TEST RESULTS-->
<!--02660656-b3d6-4311-a7d9-3e27b68f606f-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75427">build#75427</a></summary>


| test_class                    | test_method                          | test_arguments                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:------------------------------|:-------------------------------------|:--------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest     | test_write                           | null                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75427#019a3b3a-ef20-4ca8-abe2-b2e355568bfe | FLAKY         | 17/21    | upstream reliability is '92.08472686733556'. current run reliability is '80.95238095238095'. drift is 11.13235 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                            |
| ShadowLinkUpdateBrokersTests  | test_update_brokers                  | null                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75427#019a3b32-cdf2-4ed5-a5c1-549ab39f6e85 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkUpdateBrokersTests&test_method=test_update_brokers                |
| ShadowLinkUpdateBrokersTests  | test_update_brokers                  | null                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75427#019a3b3a-ef29-4bff-a37a-d150d00a038f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkUpdateBrokersTests&test_method=test_update_brokers                |
| ShadowLinkingReplicationTests | test_replication_timestamps_match    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75427#019a3b3a-ef1f-471a-8d0e-0ff16b440f75 | FLAKY         | 20/21    | upstream reliability is '95.28487229862476'. current run reliability is '95.23809523809523'. drift is 0.04678 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| DataMigrationsApiTest         | test_creating_and_listing_migrations | null                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75427#019a3b32-cdeb-4e5c-b8a2-f59135fbe1d5 | FLAKY         | 16/21    | upstream reliability is '96.52406417112299'. current run reliability is '76.19047619047619'. drift is 20.33359 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations      |
| MountUnmountIcebergTest       | test_simple_remount                  | {"cloud_storage_type": 1}                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/75427#019a3b32-cdeb-4e5c-b8a2-f59135fbe1d5 | FLAKY         | 17/21    | upstream reliability is '76.45259938837921'. current run reliability is '80.95238095238095'. drift is -4.49978 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75468">build#75468</a></summary>


| test_class                        | test_method                                  | test_arguments                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                               |
|:----------------------------------|:---------------------------------------------|:---------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade   | {"test_case_name": "All aborts"} | integration | https://buildkite.com/redpanda/redpanda/builds/75468#019a4910-02c5-4a8a-bc61-a0caa66638db | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| SegmentMsTest                     | test_segment_rolling_with_retention_consumer | null                             | integration | https://buildkite.com/redpanda/redpanda/builds/75468#019a4916-6083-4c33-96a6-2ff279412fa4 | FLAKY         | 20/21    | upstream reliability is '94.76961394769614'. current run reliability is '95.23809523809523'. drift is -0.46848 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer                   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75473">build#75473</a></summary>


| test_class                    | test_method               | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                               | test_history                                                                                                                                                          |
|:------------------------------|:--------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_auto_prefix_trimming | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75473#019a497a-514f-49d5-9ee4-5bc8aa1847cc | FLAKY         | 38/40    | upstream reliability is '92.5632911392405'. current run reliability is '90.47619047619048'. drift is 2.0871 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming |
</details>
<!--/body-->

BenPope committed:
dt/cluster_linking: Start the VerifierProducer clean

The producer does not need to continue from previous data.

This fixes a bug where stale data exists on the node:
```
More partitions in valid_offsets file than in topic!
```

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
dt/cluster_linking: Improve topic_exists_in_target

* Consistently use typing.Optional
* Use the provided rpk for checking partitions
* Simplify partition checking

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
dt/cluster_linking: Introduce topic_partitions_exists_in_target

Like topic_exists_in_target, but always wait for partitions.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
dt/cluster_linking: Use topic_partitions_exists_in_target

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ead40dc0eb3ef09b096d0829dc8f87273f8a429e..eb848e0803dfc6164d2e76d52c016cfe68371400)
* Improve `topic_exists_in_target` by using supplied rpk.
* Fix `test_update_brokers` by supplying the correct rpk.
* Start the `Producer` of `ClusterLinkingProgressVerifier` with `clean=True`

BenPope commented:
/ci-repeat 1
dt-repeat=20
release
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests

michael-redpanda commented on tests/rptest/tests/cluster_linking_test_base.py:
```diff
@@ -156,7 +156,7 @@ def start(self):
             custom_node=self.preallocated_nodes,
             **self.producer_properties,
         )
-        self.producer.start(clean=False)
+        self.producer.start(clean=True)
```
Looking at where this is used I think this is fine, though maybe providing a parameter to `def start()` would be a nice QoL but possibly not necessary at this time.
